src/enumerate-states.mc

```diff
@@ -0,0 +1,662 @@
+-- Bi-directionally maps states to integers.
+
+-- TODO(Linnea, 2022-05-30): statically resolve sizes of arrays and named upper
+-- bounds in integers
+
+include "trellis.mc"
+include "trellis-common.mc"
+
+include "mexpr/ast-builder.mc"
+include "mexpr/eval.mc"
+include "mexpr/eq.mc"
+include "option.mc"
+
+type EnumerateEnv = {
+  -- Maps a concrete type to its parameters and constructors with parameters
+  concreteTypes: Map Name ([Name], Map Name [Name]),
```

Comment:
The last `Map Name [Name]` maps constructor names to their type parameters, so for example for
```
type Option(a) {
  Some(a), None,
}
```
it would map `Some` to `[a]` and `None` to `[]`

Reply:
But couldn't you have a type like

```
type Shape {
  Rect(1..10, 1..10, 1..10, 1..10),
  Circle(1..10, 1..10, 1..10),
}
```

where some of the carried types aren't supplied through type parameters?

Reply:
Perhaps to
```
prod ConstrDecl: ConstrDecl =
  vName:UName ("(" param:TypeT ("," param:TypeT)* ")")?

prod TypeDecl : Decl =
  "type" name:UName ("(" param:LName ("," param:LName)* ")")?
  "{"
    (v:ConstrDecl ",")*
  "}"

prod TypeVar: TypeT = p:LName
```
(except that `TypeVar` introduces an LL1 conflict with `AutomatonState`)